crates/ffmpeg/src/filter_graph.rs

```diff
@@ -280,3 +280,335 @@ impl FilterContextSink<'_> {
+
+        assert!(
+            output_parser.graph.as_ptr() == filter_graph.as_ptr(),
+            "Output parser should belong to the same filter graph"
+        );
+    }
+
+    #[test]
+    fn test_filter_context_source() {
+        let mut filter_graph = FilterGraph::new().expect("Failed to create filter graph");
+        let anullsrc = Filter::get("anullsrc").expect("Failed to get 'anullsrc' filter");
+        filter_graph
+            .add(anullsrc, "src", "sample_rate=44100:channel_layout=stereo")
+            .expect("Failed to add 'anullsrc' filter");
+        let filter_context = filter_graph.get("src").expect("Failed to retrieve 'src' filter context");
+        let source_context = filter_context.source();
+
+        assert!(
+            std::ptr::eq(source_context.0, filter_graph.get("src").unwrap().0),
+            "Source context should wrap the same filter as the original filter context"
+        );
+    }
+
+    #[test]
+    fn test_filter_context_sink() {
+        let mut filter_graph = FilterGraph::new().expect("Failed to create filter graph");
+        let anullsink = Filter::get("anullsink").expect("Failed to get 'anullsink' filter");
+        filter_graph
+            .add(anullsink, "sink", "")
+            .expect("Failed to add 'anullsink' filter");
+        let filter_context = filter_graph.get("sink").expect("Failed to retrieve 'sink' filter context");
+        let sink_context = filter_context.sink();
+
+        assert!(
+            std::ptr::eq(sink_context.0, filter_graph.get("sink").unwrap().0),
+            "Sink context should wrap the same filter as the original filter context"
+        );
+    }
+
+    #[test]
+    fn test_filter_context_source_send_and_receive_frame() {
+        let mut filter_graph = FilterGraph::new().expect("Failed to create filter graph");
+        let filter_spec = "\
+            abuffer=sample_rate=44100:sample_fmt=s16:channel_layout=stereo:time_base=1/44100 \
+            [out]; \
+            [out] abuffersink";
+        FilterGraphParser::new(&mut filter_graph)
+            .parse(filter_spec)
+            .expect("Failed to parse filter graph spec");
+        filter_graph.validate().expect("Failed to validate filter graph");
+
+        let source_context_name = "Parsed_abuffer_0";
+        let sink_context_name = "Parsed_abuffersink_1";
+
+        let mut frame = Frame::new().expect("Failed to create frame");
+        frame.set_format(AV_SAMPLE_FMT_S16 as i32);
+        let mut audio_frame = frame.audio();
+        audio_frame.set_nb_samples(1024);
+        audio_frame.set_sample_rate(44100);
+
+        unsafe {
+            let av_frame = audio_frame.as_mut_ptr();
+            ffmpeg_sys_next::av_channel_layout_default(&mut (*av_frame).ch_layout, 2);
+            assert!(
+                ffmpeg_sys_next::av_frame_get_buffer(av_frame, 0) >= 0,
+                "Failed to allocate buffer for the frame"
+            );
+        }
```

Comment:
Ive seen this used a few times, perhaps we could make a method on the audio frame to do all of this so that users of the lib do not need to use unsafe.